### PR TITLE
Fix/add erc721 safe mint safe transfer from #1736

### PR DESCRIPTION
**Pull request by alanarvelo**
Hi guys, 

This addresses issue [#1736](https://github.com/OpenZeppelin/openzeppelin-solidity/issues/1736).
This is my interpretation of the solution, let me know if it looks ok. 

If it does, I will move forward implementing tests. However, before I modified any file I tried running the tests for `ERC721.test.js`.  When I ran `npm test test/token/ERC721/ERC721.test.js` I got this error:

```
1) Contract: ERC721
     "before each" hook for "returns the amount of tokens owned by the given address":
     Error: Returned error: VM Exception while processing transaction: invalid opcode
```

I noticed `ERC721.test.js` that imports a _ERC721Mock.sol_ from artifacts, which I didn't see in the contracts nor build directories. Also, running `npm test test/token/ERC721/ERC721.test.js` prompted a different error:

```
ReferenceError: e is not defined
    at async.whilst.fileFinished (/usr/local/lib/node_modules/truffle/build/webpack:/packages/truffle-compile/profiler.js:269:68)
    at next (/usr/local/lib/node_modules/truffle/build/webpack:/packages/truffle-compile/~/async/dist/async.js:5223:1)
```

which seems to come from one of the dependencies.

Any guidance on the above would be greatly appreciated. Thanks!

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -15603,7 +15584,7 @@
       "requires": {
         "underscore": "1.8.3",
         "web3-core-helpers": "1.0.0-beta.37",
-        "websocket": "git://github.com/frozeman/WebSocket-Node.git#browserifyCompatible"
+        "websocket": "git://github.com/frozeman/WebSocket-Node.git#6c72925e3f8aaaea8dc8450f97627e85263999f2"
```
Comment (frangio) [inline package-lock.json]: Where did this change come from?

File: contracts/token/ERC721/ERC721.sol [hunk 2]
```diff
@@ -174,7 +174,24 @@ contract ERC721 is ERC165, IERC721 {
...
+     * which is called upon a safe transfer, and return the magic value
+     * `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`; otherwise,
+     * the transfer is reverted.
+     * Requires the msg.sender to be the owner, approved, or operator
+     * @param from current owner of the token
+     * @param to address to receive the ownership of the given token ID
+     * @param tokenId uint256 ID of the token to be transferred
+     * @param _data bytes data to send along with a safe transfer check
+     */
+    function _safeTransferFrom(address from, address to, uint256 tokenId, bytes memory _data) public {
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: This function should be `internal`. :grimacing:

File: contracts/token/ERC721/ERC721.sol [hunk 3]
```diff
@@ -174,7 +174,24 @@ contract ERC721 is ERC165, IERC721 {
      * @param _data bytes data to send along with a safe transfer check
      */
     function safeTransferFrom(address from, address to, uint256 tokenId, bytes memory _data) public {
-        transferFrom(from, to, tokenId);
+        require(_isApprovedOrOwner(msg.sender, tokenId), "ERC721: transfer caller is not owner nor approved");
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol]: @nventuro Do we still have the guideline that every error message should be unique in the codebase?

If so this message should say something like `safe transfer`, so as to be diferent from the message in `_transferFrom`.

File: contracts/token/ERC721/ERC721.sol (hunk 2 above)
Comment (alanarvelo) [reply contracts/token/ERC721/ERC721.sol]: woops. done.

**Comment by alanarvelo** [comment]
Hi @frangio , I was able to fix my environment and run the tests. To anyone that my be reading this, I had some global packages installed that were interfering with the compilation process, after deleting them tests worked fine. 

I added your suggestions. However, changing the error message of `isApprovedOrOwner` in the `safeTransferFrom` function, causes problems down the line as this modifier is also used in `trasnferFrom` function and they are tested together in the codebase. 

Such test is in `ERC721Behavior.js` under the function `shouldTransferTokensByUsers`:

**Comment by frangio** [comment]
Hm, yes that is complex. Let's revert to the previous error message for now. (cc @nventuro)

Coverage decreased because the new `_safeMint` function is not tested yet. [See the report here.](https://coveralls.io/builds/24558651/source?filename=contracts/token/ERC721/ERC721.sol#L231)

Since it's an internal function you will have to add it to `ERC721Mock`, and then add a new test for it in `ERC721.behavior.js`.

**Comment by alanarvelo** [comment]
Hello @frangio, 
Added `_safeMint` functions to `ERC721Mock.sol` and created tests — similar to those that check the `_checkOnERC721Received` functionality of the `safeTransferFrom` functions — to check the now the `_checkOnERC721Received` of the safeMint functions.

Regular minting functionality is tested in other test files for the `_mint` function, which is what the `safeMint` functions use internally, so didn't test that.

How does it look?

**Comment by frangio** [comment]
This is in a great place @alanarvelo!

> Some tests are failing, apparently because other ERC721 versions (Pausable, Burnable, etc) are not inheriting the safeMint functions. Any ideas?

I don't see any failing tests.

> Also, wouldn't it make sense to add a new tests for safeMint in ERC721MintBurn.behavior.js?

`ERC721MintBurn` tests the public interface, and this PR so far only introduces an internal function. However, now that you mention it, we should add a public `safeMint` function in `ERC721Mintable`, and then we would have to test it as well. Can you add that?

**Comment by alanarvelo** [comment]
Hi @frangio I made another PR right after that fixed the failing tests, careful that you are not looking at the older one, I restructured the way `safeMint` is tested by focusing only on `ERC721Mock`.

I will add that public `safeMint` fn to `ERC721Mintable` and add a test for it. Where should that test go? I assume in `ERC721MintBurn.behavior.js` as ERC721Mintable.test.js` just calls the `shouldBehaveLikeMintAndBurnERC721` exported from the former. Am I correct?

**Comment by frangio** [comment]
>Where should that test go? I assume in `ERC721MintBurn.behavior.js`

Yeah, exactly!

**Comment by alanarvelo** [comment]
Hi @frangio added `safeMint` fns to `ERC721Mintable.sol` as agreed, and corresponding tests. 
Let me know your thoughts.

**Comment by alanarvelo** [comment]
Awesome 😃 
No problem, thank you!

**Pull request merged**